src/core/core.cpp

```diff
@@ -766,59 +764,61 @@ wayfire_view wf::compositor_core_impl_t::find_view(const std::string& id)
 
 pid_t wf::compositor_core_impl_t::run(std::string command)
 {
-    static constexpr size_t READ_END  = 0;
-    static constexpr size_t WRITE_END = 1;
     pid_t pid;
-    int pipe_fd[2];
-    pipe2(pipe_fd, O_CLOEXEC);
+    int daemonized, status, sp[2];
+    struct pollfd pfd;
+    int TIMEOUT = 500;
+
+    if (socketpair(AF_UNIX, SOCK_STREAM, PF_UNSPEC, sp) == -1)
+    {
+      return -1;
+    }
 
-    /* The following is a "hack" for disowning the child processes,
-     * otherwise they will simply stay as zombie processes */
     pid = fork();
+    if (pid < 0)
+    {
+        return -1;
+    }
     if (!pid)
     {
-        pid = fork();
-        if (!pid)
+        daemonized = daemon(1, 0);
+        if (daemonized != 0)
         {
-            close(pipe_fd[READ_END]);
-            close(pipe_fd[WRITE_END]);
-
-            setenv("_JAVA_AWT_WM_NONREPARENTING", "1", 1);
-            setenv("WAYLAND_DISPLAY", wayland_display.c_str(), 1);
+            _exit(daemonized);
+        }
+        /* return daemon pid back to wayfire */
+        pid = getpid();
+        close(sp[0]);
+        write(sp[1], (void *)&pid, sizeof(pid));
+        close(sp[1]);
+
+        setenv("_JAVA_AWT_WM_NONREPARENTING", "1", 1);
+        setenv("WAYLAND_DISPLAY", wayland_display.c_str(), 1);
 #if WF_HAS_XWAYLAND
-            if (!xwayland_get_display().empty())
-            {
-                setenv("DISPLAY", xwayland_get_display().c_str(), 1);
-            }
-
-#endif
-            int dev_null = open("/dev/null", O_WRONLY);
-            dup2(dev_null, 1);
-            dup2(dev_null, 2);
-            close(dev_null);
-
-            _exit(execl("/bin/sh", "/bin/sh", "-c", command.c_str(), NULL));
-        } else
+        if (!xwayland_get_display().empty())
         {
-            close(pipe_fd[READ_END]);
-            write(pipe_fd[WRITE_END], (void*)(&pid), sizeof(pid));
-            close(pipe_fd[WRITE_END]);
-            _exit(0);
+            setenv("DISPLAY", xwayland_get_display().c_str(), 1);
         }
-    } else
-    {
-        close(pipe_fd[WRITE_END]);
-
-        int status;
-        waitpid(pid, &status, 0);
-
-        pid_t child_pid;
-        read(pipe_fd[READ_END], &child_pid, sizeof(child_pid));
+#endif
+        execl("/bin/sh", "/bin/sh", "-c", command.c_str(), NULL);
```

Comment:
I see you have removed the `_exit()` call. Wouldn't that cause issues if `execl` fails?

Reply:
I still do have the `_exit` but it's moved above to line 787.

Reply:
Yeah but the exit should happen after a failed exec